nemo/collections/tts/models/radtts.py

```diff
@@ -0,0 +1,419 @@
+        )  # batch['speaker_id'] single speaker for now
+        text = batch['text']  # batch['text']
+        in_lens = batch['text_lens']
+        out_lens = batch['log_mel_lens']
+        attn_prior = batch['align_prior_matrix']
+        f0 = batch['pitch']
+        voiced_mask = batch['voiced_mask']
+        p_voiced = batch['p_voiced']
+        energy_avg = batch['energy']
+        mel = batch['log_mel']
+        if (
+            self.train_config.binarization_start_iter >= 0
+            and self.global_step >= self.train_config.binarization_start_iter
+        ):
+            # binarization training phase
+            binarize = True
+        else:
+            # no binarization, soft-only
+            binarize = False
+        outputs = self.model(
+            mel,
+            speaker_ids,
+            text,
+            in_lens,
+            out_lens,
+            binarize_attention=True,
+            attn_prior=attn_prior,
+            f0=f0,
+            energy_avg=energy_avg,
+            voiced_mask=voiced_mask,
+            p_voiced=p_voiced,
+        )
+        loss_outputs = self.criterion(outputs, in_lens, out_lens)
+
+        loss = None
+        for k, (v, w) in loss_outputs.items():
+            if w > 0:
+                loss = v * w if loss is None else loss + v * w
+
+        if (
+            binarize
+            and self.train_config.kl_loss_start_iter >= 0
+            and self.global_step >= self.train_config.kl_loss_start_iter
+        ):
+            binarization_loss = self.attention_kl_loss(outputs['attn'], outputs['attn_soft'])
+            loss += binarization_loss
+        else:
+            binarization_loss = torch.zeros_like(loss)
+        loss_outputs['binarization_loss'] = binarization_loss
+
+        return {
+            "loss_outputs": loss_outputs,
+            "attn": outputs["attn"] if batch_idx == 0 else None,
+            "attn_soft": outputs["attn_soft"] if batch_idx == 0 else None,
+            "audiopaths": "audio_1" if batch_idx == 0 else None,
+        }
+
+    def validation_epoch_end(self, outputs):
+
+        loss_outputs = outputs[0]["loss_outputs"]
+
+        for k, v in loss_outputs.items():
+            if k != "binarization_loss":
+                self.log("val/" + k, loss_outputs[k][0])
+
+        attn = outputs[0]["attn"]
+        attn_soft = outputs[0]["attn_soft"]
+
+        self.tb_logger.add_image(
+            'attention_weights_mas',
+            plot_alignment_to_numpy(attn[0, 0].data.cpu().numpy().T, title="audio"),
+            self.global_step,
+            dataformats='HWC',
+        )
+
+        self.tb_logger.add_image(
+            'attention_weights',
+            plot_alignment_to_numpy(attn_soft[0, 0].data.cpu().numpy().T, title="audio"),
+            self.global_step,
+            dataformats='HWC',
+        )
+
+        self.log_train_images = True
+
+    def configure_optimizers(self):
```

Comment:
This is built into `ModelPT` and you should try to leverage that instead of redefining it

Reply:
We are using Radam optimizer (which I don't think it's built into ModelPT optimizers). So I needed to initialize and pass it.